### PR TITLE
[Epic] Anticheat runtimes (EAC and BattlEye)

### PR DESCRIPTION
**Pull request by CommandMC**
It is now possible to manage and enable anticheat runtimes in the launcher
Current issues:
- ~~BattlEye runtime doesn't work, the Wine-GE patch for it seems to not apply. Not sure why, needs more testing~~
   Nevermind, the runtime works, the only game I was testing with (The Cycle: Frontier) did not work correctly

---

Use the following Checklist if you have changed something on the Backend or Frontend:

- [X] Tested the feature and it's working on a current and clean install.
- [X] Tested the main App features and they are still working on a current and clean install. (Login, Install, Play, Uninstall, Move games, etc.)
- [ ] Created / Updated Tests (If necessary)
- [ ] Created / Updated documentation (If necessary)

### CODE REVIEW

File: electron/wine/runtimes/ipc_handler.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { RuntimeName } from './../../types'
+import { ipcMain } from 'electron'
+import { download, isInstalled } from './runtimes'
+
+ipcMain.handle('downloadRuntime', async (e, runtime_name: RuntimeName) => {
+  return download(runtime_name)
+})
```
Comment (Nocccer) [inline electron/wine/runtimes/ipc_handler.ts:7]:
```suggestion
})

```

File: electron/wine/runtimes/runtimes.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+import axios from 'axios'
+import { existsSync, mkdirSync, unlinkSync, writeFileSync } from 'graceful-fs'
+import { join } from 'path'
+import { runtimePath } from './../../constants'
+import { logDebug, LogPrefix } from './../../logger/logger'
+import { Runtime, RuntimeName } from './../../types'
+import { extractTarFile, getAssetDataFromDownload } from './util'
+
+async function _get(): Promise<Runtime[]> {
+  mkdirSync(runtimePath, { recursive: true })
```
Comment (Nocccer) [inline electron/wine/runtimes/runtimes.ts:10]: Please catch possible errors here and log them!

File: electron/wine/runtimes/runtimes.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  const allRuntimes = await axios.get('https://lutris.net/api/runtimes')
+  return allRuntimes.data
+}
+
+async function download(name: RuntimeName) {
+  const runtimes = await _get()
+  const runtime = runtimes.find((inst) => {
```
Comment (Nocccer) [inline electron/wine/runtimes/runtimes.ts:17]: Check if runtimes is not undefined, else you will get trace with `find is not a function of undefined`

File: electron/wine/runtimes/runtimes.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    return inst.name === name
+  })
+  logDebug(
```
Comment (Nocccer) [inline electron/wine/runtimes/runtimes.ts:20]: Would go with logInfo here. Your choice.

File: electron/wine/runtimes/runtimes.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    LogPrefix.Runtime
+  )
+  const { name: tarFileName, content_type } = await getAssetDataFromDownload(
+    runtime.url
+  )
+  const rawData = (
+    await axios.get(runtime.url, { responseType: 'arraybuffer' })
+  ).data
+  const tarFilePath = join(runtimePath, tarFileName)
+  writeFileSync(tarFilePath, rawData)
```
Comment (Nocccer) [inline electron/wine/runtimes/runtimes.ts:31]: Catch possible errors!

File: electron/wine/runtimes/runtimes.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+  const extractedFolderPath = join(runtimePath, name)
+  await extractTarFile(tarFilePath, content_type, extractedFolderPath)
+  unlinkSync(tarFilePath)
```
Comment (Nocccer) [inline electron/wine/runtimes/runtimes.ts:34]: Catch possible errors!

File: electron/wine/runtimes/util.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+): Promise<GithubAssetMetadata> {
+  const splitUrl = url.split('/').filter(Boolean)
+  const [, , author, repo, , , tag, assetName] = splitUrl
+  const releaseData = (
+    await axios.get(
+      `https://api.github.com/repos/${author}/${repo}/releases/tags/${tag}`
+    )
+  ).data
+  const assets: Array<GithubAssetMetadata> = releaseData.assets
+  return assets.find((asset) => {
```
Comment (Nocccer) [inline electron/wine/runtimes/util.ts:37]: Check that assets is not undefined, else you can get `find is not a function of undefined`

File: electron/wine/runtimes/util.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    if (asset.name === assetName) {
+      return true
+    }
+  })
+}
+
+async function downloadFile(url: string, filePath: string) {
+  const data = (await axios.get(url)).data
+  writeFileSync(filePath, data)
```
Comment (Nocccer) [inline electron/wine/runtimes/util.ts:46]: Catch errors. What if axios fails to download and what if writeFileSync fails , too?

File: electron/wine/runtimes/util.ts [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+  filePath: string,
+  contentType: string,
+  extractedPath?: string
+) {
+  if (!extractedPath) {
+    const splitPath = filePath.split('.tar')
+    splitPath.pop()
+    extractedPath = splitPath.join('.tar')
+  }
+  mkdirSync(extractedPath, { recursive: true })
```
Comment (Nocccer) [inline electron/wine/runtimes/util.ts:59]: catch possible errors

File: src/screens/Settings/components/WineExtensions/index.tsx [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+        }
+      }
+
+      const isInstalled = await ipcRenderer.invoke(
+        'isRuntimeInstalled',
+        'eac_runtime'
+      )
+      if (!isInstalled) {
+        setEacInstalling(true)
+        await ipcRenderer.invoke('downloadRuntime', 'eac_runtime')
```
Comment (Nocccer) [inline src/screens/Settings/components/WineExtensions/index.tsx:56]: If you work with reject/ resolve you should catch and print out errors. Or you catch the errors directly in the backend.

File: src/screens/Settings/components/WineExtensions/index.tsx [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+
+  const handleBattlEyeRuntime = async () => {
+    if (!battlEyeRuntime) {
+      const isInstalled = await ipcRenderer.invoke(
+        'isRuntimeInstalled',
+        'battleye_runtime'
+      )
+      if (!isInstalled) {
+        setBattlEyeInstalling(true)
+        await ipcRenderer.invoke('downloadRuntime', 'battleye_runtime')
```
Comment (Nocccer) [inline src/screens/Settings/components/WineExtensions/index.tsx:70]: If you work with reject/ resolve you should catch and print out errors. Or you catch the errors directly in the backend.

**Review by Nocccer — CHANGES_REQUESTED** [review]
Beside the error evaluation, this is a clean code.
But the error handling is important, so need to reject it.
How you handle the errors is your choice. You can do it in the backend or in the frontend.

Btw you could also write backend tests. Downloading/unzipping etc. You can have a look at heroic-wine-downloader tests, how i simulate a url request, download etc.

**Comment by CommandMC** [comment]
Wasn't done with this yet, otherwise I would've marked it as ready for review

I understand your concerns and will implement error handling in the new functions. As for tests, do you have a piece of documentation or something I can look at, to make sure I'm doing it right?

**Comment by Nocccer** [comment]
> Wasn't done with this yet, otherwise I would've marked it as ready for review
> 
> I understand your concerns and will implement error handling in the new functions. As for tests, other than examples, do you have a piece of documentation or something I can look at, to make sure I'm doing it right?

Ah sorry didn't saw the wip. Ok i will resolve all comments for now.
Jest documentation: https://jestjs.io/docs/getting-started
Everything is setup already. You just need to create a new *.test.ts and write test's there.
If you need help i can try to write the first test for you as an example.

File: electron/wine/runtimes/runtimes.ts (hunk 2 above)
Comment (CommandMC) [reply electron/wine/runtimes/runtimes.ts:10]: We're using mkdirSync without error handling in [a](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/blob/main/electron/config.ts#L220) [lot](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/blob/main/electron/launcher.ts#L353) [of](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/blob/main/electron/main.ts#L1177) [other](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/blob/main/electron/main.ts#L1177) [places](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/blob/main/electron/main.ts#L1177) as well

**Comment by CommandMC** [comment]
Alright, error handling is now in place. Let me know what you think, I'll take a look at tests now then

File: electron/__tests__/runtimes/utils.test.ts [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+
+    await expect(downloadFile(testUrl, '')).rejects.toEqual(
+      Error('Failed to save downloaded data to file: Mocked error stack')
+    )
+  })
+})
+
+describe('extractTarFile', () => {
+  it('Success without strip', async () => {
+    const tmpDir = dirSync()
```
Comment (Nocccer) [inline electron/__tests__/runtimes/utils.test.ts:155]: This variable needs to be true, else tmp can't remove none empty folders.
```suggestion
    const tmpDir = dirSync({unsafeCleanup: true})
```

Doc:
```typescript
    /**
     * Recursively removes the created temporary directory, even when it's not empty.
     * @default false
     */
    unsafeCleanup?: boolean | undefined;
```

**Comment by CommandMC** [comment]
There are still a lot of tests missing, although the functionality itself might be nice to have for a beta release

**Pull request merged**